### PR TITLE
Add InterceptorBase

### PR DESCRIPTION
**Pull request by isdaniel**
Add an abstract class and implement Interceptor interface, other class can use it like Asp.net MVC action filter.

### CODE REVIEW

File: src/Castle.Core/DynamicProxy/IInterceptorBase.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
+﻿
```
Comment (stakx) [inline src/Castle.Core/DynamicProxy/IInterceptorBase.cs]: Copyright notice is missing, see other source code files.

File: src/Castle.Core/DynamicProxy/IInterceptorBase.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+namespace Castle.DynamicProxy
+{
+	using System;
+
+	public class ExceptionContext
+	{
+
```
Comment (stakx) [inline src/Castle.Core/DynamicProxy/IInterceptorBase.cs]: Superfluous blank line, should be removed.

File: src/Castle.Core/DynamicProxy/IInterceptorBase.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+		public Exception Ex { get; set; }
+		public IInvocation Invocation { get; set; }
+	}
+
+	public abstract class InterceptorBase : IInterceptor
+	{
+		protected virtual void OnExcuting(IInvocation invocation)
```
Comment (stakx) [inline src/Castle.Core/DynamicProxy/IInterceptorBase.cs]: `OnExcuting` &rarr; `OnExecuting`.

File: src/Castle.Core/DynamicProxy/IInterceptorBase.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+		{
+		}
+
+		protected virtual void OnExcuted(IInvocation invocation)
```
Comment (stakx) [inline src/Castle.Core/DynamicProxy/IInterceptorBase.cs]: `OnExcuted` &rarr; `OnExecuted`.

File: src/Castle.Core/DynamicProxy/IInterceptorBase.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+		}
+
+		protected virtual void OnException(ExceptionContext context)
+		{
+		}
+
+		private bool NotVoidHasValue(IInvocation invocation)
+		{
+			return !(invocation.Method.ReturnType == typeof(void) ||
+			         invocation.ReturnValue == null);
```
Comment (stakx) [inline src/Castle.Core/DynamicProxy/IInterceptorBase.cs]: What if a non-`void` method *wants to* return `null`? It's a perfectly valid value, after all.

File: src/Castle.Core/DynamicProxy/IInterceptorBase.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+		public void Intercept(IInvocation invocation)
+		{
+			try
+			{
+				OnExcuting(invocation);
+
+				if (NotVoidHasValue(invocation))
+					return;
+
+				invocation.Proceed();
```
Comment (stakx) [inline src/Castle.Core/DynamicProxy/IInterceptorBase.cs]: I don't think that this pattern&mdash;`Proceed` if the interceptor hasn't yet provided a return value&mdash;holds true for all use cases (though admittedly for a good many). Sometimes there might be nothing to proceed *to*. You'd at least have to document your interceptor's behavior / functional semantics.

File: src/Castle.Core/DynamicProxy/IInterceptorBase.cs [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+		{
+			return !(invocation.Method.ReturnType == typeof(void) ||
+			         invocation.ReturnValue == null);
+		}
+
+		public void Intercept(IInvocation invocation)
+		{
+			try
+			{
+				OnExcuting(invocation);
```
Comment (stakx) [inline src/Castle.Core/DynamicProxy/IInterceptorBase.cs]: `OnExcuting` &rarr; `OnExecuting`.

File: src/Castle.Core/DynamicProxy/IInterceptorBase.cs [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+				if (NotVoidHasValue(invocation))
+					return;
+
+				invocation.Proceed();
+
+				OnExcuting(invocation);
```
Comment (stakx) [inline src/Castle.Core/DynamicProxy/IInterceptorBase.cs]: `OnExcuting` &rarr; `OnExecuted`.

File: src/Castle.Core/DynamicProxy/IInterceptorBase.cs [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+	using System;
+
+	public class ExceptionContext
+	{
+
+		public Exception Ex { get; set; }
+		public IInvocation Invocation { get; set; }
+	}
+
+	public abstract class InterceptorBase : IInterceptor
```
Comment (stakx) [inline src/Castle.Core/DynamicProxy/IInterceptorBase.cs]: I find your interceptor to actually have a fairly specific logic (which unfortunately isn't documented at all e.g. in XML doc comments), and for that this name seems to generic. And there's my main issue with this PR: The interceptor seems at the same time too specific for a low-level infrastructure library like DynamicProxy, and short enough to be easily written in your own application code.

File: src/Castle.Core/DynamicProxy/IInterceptorBase.cs [hunk 10]
```diff
@@ -0,0 +1,56 @@
+﻿
+namespace Castle.DynamicProxy
+{
+	using System;
+
+	public class ExceptionContext
```
Comment (stakx) [inline src/Castle.Core/DynamicProxy/IInterceptorBase.cs]: This should be `sealed` IMO (as there's no reason why someone outside the library should be able to instantiate this type), or even better, a `readonly struct` (as it's essentially a 2-tuple).

Also, documentation never hurts.

File: src/Castle.Core/DynamicProxy/IInterceptorBase.cs (hunk 5 above)
Comment (isdaniel) [reply src/Castle.Core/DynamicProxy/IInterceptorBase.cs]: Hi @stakx 
In my opinion
`invocation.ReturnValue` initial is `null`, developer set `invocation.ReturnValue`  that mean we don't want to execute this method (like do authorize).

thank you.

**Comment by isdaniel** [comment]
@jonorossi @stakx  Thank for your comment and helpful suggestions.
I benefited a lot from your advice.

**Pull request closed without merging**